go/pools/smartconnpool/pool.go

```diff
@@ -430,18 +430,40 @@ func (pool *ConnPool[C]) tryReturnConn(conn *Pooled[C]) bool {
 	if pool.wait.tryReturnConn(conn) {
 		return true
 	}
-	if pool.closeOnIdleLimitReached(conn) {
+
+	for {
+		if pool.capacity.Load() < pool.active.Load() {
+			conn.Close()
+			pool.closedConn()
+			return true
+		}
+
+		if pool.closeOnIdleLimitReached(conn) {
+			return false
+		}
+
+		if !pool.capacityMu.TryRLock() {
+			// If we can't get a read lock here, it means that the pool is being closed. Retry and check `capacity` again.
+			continue
+		}
+		defer pool.capacityMu.RUnlock()
+
+		if pool.capacity.Load() < pool.active.Load() {
+			conn.Close()
+			pool.closedConn()
+			return true
+		}
+
+		connSetting := conn.Conn.Setting()
+		if connSetting == nil {
+			pool.clean.Push(conn)
+		} else {
+			stack := connSetting.bucket & stackMask
+			pool.settings[stack].Push(conn)
+			pool.freshSettingsStack.Store(int64(stack))
```

Comment:
I wonder, would an alternative approach work to avoid having to lock? What if we still optimistically put it into the stack, but then check after we've completed that if we have since been closed and then remove it again? 

So you only pay the price during closing mostly and then having to do any locking or other synchronization?

Reply:
I'm definitely open to alternative approaches. For your suggestion,  don't we have to take a lock on `capacityMu` to correctly decide whether the pool is closed or not?

Reply:
I am working on an additional change that uses a pool generation number to allow `Close()` to return instantly and basically break the link between checked out connections and the pool.

I think that could be used for the optimistic concurrency check. I'll try this out and see if I can make this work.